proposals/002-text-utf-default.md

```diff
@@ -0,0 +1,156 @@
+This proposal is for the migration of the `text` package from its
+default UTF-16 encoding to UTF-8. The lack of UTF-8 as a default in the
+`text` package is a pain point raised by the Haskell Community and
+industry partners for many years. The Haskell Foundation is uniquely
+positioned to effect a change like this, granted there is an appetite
+for breakage, and that such a project is well-socialized and planned.
+During the meetings of the Haskell Foundation Tech Agenda Track, we
+identified the pros and cons of such a migration against historical
+attempts, and agreed that it was within our power to do.
+
+Further, we solicited both community and industry feedback to gauge the
+appetite for breakage, and what stakeholders would be affected by the
+changes. Andrew Lelechenko has offered to lead the implementation of
+this project.
+
+# Motivation
+
+-   UTF-16 by default requires that all Text values pay a premium for serialization. Arguably, the performance impact of Text is flipped
+    upside-down: most text is UTF-8, and Haskell devs pay an undue cost when working with the wrong default.
+
+-   UTF-8 is the industry standard and by far the most common text encoding, with roughly 97% of web pages existing in UTF-8. The
+    existing UTF-16 default imposes an additional hurdle to working with the vast majority of web content on earth.
+
+-   Many systems in Haskell are UTF-8 by default (e.g. Haddock)
+
+# Goals
+
+-   Solicit feedback from community members and industry to gauge appetite for such a change.
+
+-   Provide an implementation, migration, and delivery plan for changing the default encoding of `text` from UTF-16 to UTF-8.
+
+-   Ensure stakeholders (e.g. GHC, Cabal, Stack, boot libs) have ample time to migrate and address any bugs.
+
+-   Implementation should not significantly alter the performance characteristics of the base `text` library within some tolerance
+    threshold.
+
```

Comment:
Are there any API stability goals? I know this is mentioned elsewhere, but it would be nice to define here that, say, a concrete goal is that all the functions exported from `Data.Text` remain unchanged in their types and semantics (but may change in performance characteristics). I have no idea whether that's the right goal, but if stability is a goal of the project, this is the place to put it.

Reply:
This would be a good deliverable actually. The API *must* remain the same in order for migration to be as seamless as possible. Since we're building on `text-utf8`, this is actually already done for us for the most part, and we intend to keep it confluent with the existing `text` API. I'll add this.

Reply:
Yes, but then we also need to define what we mean by "the `text` API".

Reply:
That's a very good point. This is an implementation detail that we'll need to solve, and I believe you're hinting at whether the internals should be included in the API.

Reply:
I've added a section describing API compatibility in more details.